images/virtualization-artifact/pkg/controller/vmip/internal/lifecycle_handler.go

```diff
@@ -0,0 +1,139 @@
+	vmip := state.VirtualMachineIP()
+	vmipStatus := &vmip.Status
+
+	vm, err := state.VirtualMachine(ctx)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	mgr := conditions.NewManager(vmipStatus.Conditions)
+	conditionBound := conditions.NewConditionBuilder(vmipcondition.BoundType).
+		Generation(vmip.GetGeneration())
+
+	conditionAttach := conditions.NewConditionBuilder(vmipcondition.AttachedType).
+		Generation(vmip.GetGeneration())
+
+	if vm != nil {
+		vmipStatus.VirtualMachine = vm.Name
+		mgr.Update(conditionAttach.Status(metav1.ConditionTrue).
+			Reason(vmipcondition.Attached).
+			Condition())
+	} else {
+		vmipStatus.VirtualMachine = ""
+		mgr.Update(conditionAttach.Status(metav1.ConditionFalse).
+			Reason(vmipcondition.VirtualMachineNotFound).
+			Message("Virtual machine not found").
+			Condition())
+	}
+
+	lease, err := state.VirtualMachineIPLease(ctx)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	switch {
+	case lease == nil && vmipStatus.Address != "":
+		if vmipStatus.Phase != virtv2.VirtualMachineIPAddressPhasePending {
+			vmipStatus.Phase = virtv2.VirtualMachineIPAddressPhasePending
+			mgr.Update(conditionBound.Status(metav1.ConditionFalse).
+				Reason(vmipcondition.VirtualMachineIPAddressLeaseLost).
+				Message(fmt.Sprintf("VirtualMachineIPAddressLease %s doesn't exist",
+					common.IpToLeaseName(vmipStatus.Address))).
+				Condition())
+		}
+
+	case lease == nil:
+		if vmipStatus.Phase != virtv2.VirtualMachineIPAddressPhasePending {
+			vmipStatus.Phase = virtv2.VirtualMachineIPAddressPhasePending
+			mgr.Update(conditionBound.Status(metav1.ConditionFalse).
+				Reason(vmipcondition.VirtualMachineIPAddressLeaseNotFound).
+				Message("VirtualMachineIPAddressLease not found").
+				Condition())
+		}
+
+	case util.IsBoundLease(lease, vmip):
+		if vmipStatus.Phase != virtv2.VirtualMachineIPAddressPhaseBound {
+			vmipStatus.Phase = virtv2.VirtualMachineIPAddressPhaseBound
+			vmipStatus.Address = common.LeaseNameToIP(lease.Name)
+			mgr.Update(conditionBound.Status(metav1.ConditionTrue).
+				Reason(vmipcondition.Bound).
+				Condition())
+		}
+
+	case lease.Status.Phase == virtv2.VirtualMachineIPAddressLeasePhaseBound:
+		if vmipStatus.Phase != virtv2.VirtualMachineIPAddressPhasePending {
+			vmipStatus.Phase = virtv2.VirtualMachineIPAddressPhasePending
+			mgr.Update(conditionBound.Status(metav1.ConditionFalse).
+				Reason(vmipcondition.VirtualMachineIPAddressLeaseAlready).
+				Message(fmt.Sprintf("VirtualMachineIPAddressLease %s is bound to another VirtualMachineIP",
+					common.IpToLeaseName(vmipStatus.Address))).
+				Condition())
+		}
+
+	default:
+		if vmipStatus.Phase != virtv2.VirtualMachineIPAddressPhasePending {
+			vmipStatus.Phase = virtv2.VirtualMachineIPAddressPhasePending
+			mgr.Update(conditionBound.Status(metav1.ConditionFalse).
+				Reason(vmipcondition.VirtualMachineIPAddressLeaseNotFound).
```

Comment:
Why not found? It is found

Reply:
Yes, it is state when lease not ready. Added new condition 'VirtualMachineIPAddressLeaseNotReady'